src/components/attendee/VideoComms/internal/TwilioImplementation.tsx

```diff
@@ -0,0 +1,396 @@
+      recalculateStatus();
+    };
+  };
+
+  const recalculateStatus = () => {
+    // This isn't particularly performant. However, we generally don't get
+    // a lot of state changes happening during a call so this is probably ok.
+    // It's more important to present a simple API than it is to overly fixate
+    // on performance
+    const localParticipant = updateLocalParticipant();
+    const remoteParticipants = Array.from(
+      room?.participants.values() || []
+    ).map(wrapRemoteParticipant);
+
+    onStateUpdateCallback({
+      localParticipant,
+      status,
+      remoteParticipants,
+      isTransmittingAudio,
+      isTransmittingVideo,
+    });
+  };
+
+  const startAudio = async () => {
+    if (!room) {
+      console.warn("startAudio called from invalid state");
+      return;
+    }
+    if (!room.localParticipant.audioTracks.size) {
+      // This situation arises when the local participant joins a call without
+      // audio enabled at the start. Create the audio stream now.
+      // Update the UI whilst we're publishing the streams
+      isTransmittingAudio = true;
+      recalculateStatus();
+
+      const localAudioTrack = await Twilio.createLocalAudioTrack();
+      await room.localParticipant.publishTrack(localAudioTrack);
+      recalculateStatus();
+    }
+    for (const track of room.localParticipant.audioTracks.values()) {
+      track.track.enable();
+    }
+    isTransmittingAudio = true;
+    recalculateStatus();
+  };
+  const stopAudio = () => {
+    if (!room) {
+      console.warn("stopAudio called from invalid state");
+      return;
+    }
+    for (const track of room.localParticipant.audioTracks.values()) {
+      track.track.disable();
+    }
+    isTransmittingAudio = false;
+    recalculateStatus();
+  };
+  const startVideo = async () => {
+    if (!room) {
+      console.warn("startVideo called from invalid state");
+      return;
+    }
+    if (!room.localParticipant.videoTracks.size) {
+      // This situation arises when the local participant joins a call without
+      // video enabled at the start. Create the video stream now.
+      isTransmittingVideo = true;
+      recalculateStatus();
+
+      const localVideoTrack = await Twilio.createLocalVideoTrack();
+      await room.localParticipant.publishTrack(localVideoTrack);
+      recalculateStatus();
+    }
+    for (const track of room.localParticipant.videoTracks.values()) {
+      track.track.enable();
+    }
+    isTransmittingVideo = true;
+    recalculateStatus();
+  };
+  const stopVideo = () => {
+    if (!room) {
+      console.warn("stopVideo called from invalid state");
+      return;
+    }
+    for (const track of room.localParticipant.videoTracks.values()) {
+      track.track.disable();
+    }
+    isTransmittingVideo = false;
+    recalculateStatus();
+  };
+
+  return {
+    joinChannel,
+    disconnect,
+    shareScreen,
+    startAudio,
+    stopAudio,
+    startVideo,
+    stopVideo,
+  };
```

Comment:
That's my next possible refactor. When I first built this is maintained more internal state.. Then I simplified it right down to re-generating state as that led to fewer bugs ;)

Reply:
@mike-lvov are we thinking of the same definition of "pure function"?

https://en.wikipedia.org/wiki/Pure_function

I'm considering the "react way" of at least having side effects  (`like a Reac.FC` is pure), but having the functions depend on explicit input, something like an object that has the state/context outside so that we can add/remove functions in the future directly at the file top level, not being tied to a closure.

I'm considering any hook that calls those functions will provide the closure (have the state/context object)

Reply:
> I doubt it would make this code more readable

Readability is one issue.

Being able to easily test something is another. The more predictable a function is (like one that doesn't depend in implicit input), the easier is to give it value, check the result, confirm it works or doesn't.

Reply:
I started moving these functions out into top-level functions. The number of arguments that would need passing in would make them worse in my opinion. I might come back to this later and see what can be done to make this easier to work with.

Reply:
Number of arguments is 1. Just make a type for it